### PR TITLE
Collision improvement

### PR DESCRIPTION
**Pull request by gabrieldulca**
Single selection element jumps over collision targets and element snapping according to #309 .

### CODE REVIEW

File: client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts [hunk 1]
```diff
@@ -148,6 +151,132 @@ export class FeedbackMoveMouseListener extends MouseListener {
...
+        };
+        const elementBottomLeft: Point = {
+            x: element.bounds.x,
+            y: element.bounds.y + element.bounds.height
+        };
+        const elementBottomRight: Point = {
+            x: element.bounds.x + element.bounds.width,
+            y: element.bounds.y + element.bounds.height
+        };
+
```
Comment (planger) [inline client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts]: Please remove multiple consecutive blank lines

File: client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts [hunk 2]
```diff
@@ -148,6 +151,132 @@ export class FeedbackMoveMouseListener extends MouseListener {
...
+        const numerator: number = Math.abs((secondLine.a * p1.x) + (secondLine.b * p1.y) - secondLine.c);
+        const denominator: number = Math.sqrt(Math.pow(secondLine.a, 2) + Math.pow(secondLine.b, 2));
+        return numerator / denominator;
+    }
+
+    /**
+     * Snaps the element to the target in case of a collision
+     */
+    getSnappedBounds(element: SModelElement & BoundsAware, target: SModelElement & BoundsAware): Bounds {
+        let snappedBounds: Bounds = element.bounds;
```
Comment (planger) [inline client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts]: Please avoid specifying the type when it is actually inferred (line `snappedBounds: Bounds`). There are several occurrences of unnecessary type specifications. Please remove them all to make the code more concise.

File: client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts [hunk 3]
```diff
@@ -170,18 +299,64 @@ export class FeedbackMoveMouseListener extends MouseListener {
                     if (isMoveable(element) && isResizeable(element)) {
                         // If noElementOverlap Option is set perform collision detection
                         if (this.glspViewerOptions.noElementOverlap) {
+
```
Comment (planger) [inline client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts]: I find this very long block of code inside of a inline forEach very hard to read and mouseMove(...) is already a very long and technically deep method. I think it would be good to factor it out. After all, it handles a case which is optional (noElementOverlap configured), so it is a separate concern and should be in an own method/function. Also, I have the feeling this would be easy to do, since eventually we only add actions to the result, so the extracted methods could return actions if they want to, and here we only collect them.

File: client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts [hunk 4]
```diff
@@ -170,18 +299,64 @@ export class FeedbackMoveMouseListener extends MouseListener {
...
+                                                },
+                                                toPosition: {
+                                                    x: snappedBounds.x,
+                                                    y: snappedBounds.y
+                                                }
+                                            });
+                                            result.push(new MoveAction(snapMoves, false));
+                                        }
+
+
```
Comment (planger) [inline client/packages/sprotty-client/src/features/tool-feedback/change-bounds-tool-feedback.ts]: Sorry, github doesn't allow me to comment on code that hasn't changed. So this comment is about the lines below.
In line 368 and 369, there is a very complex condition that is hard to follow.
Also, is it really necessary to access `willCollide` and `hasCollided` down there, in a block that is not inside the code for the case where `noElementOverlap` is on? I find it a bit confusing to leak those variables into the "I don't care about collisions" case.
If a state has to leak, the switch should be very general, e.g. `isMoveAllowed`.
Lastly, couldn't we make the terminology a bit more consistent wrt element overlap and collision. We have the switch `noElementOverlap`, why don't we name the `willCollide` also `willOverlap`?

**Comment by tortmayr** [comment]
> Finally, I think it would be better to squash this confusing change history into a single commit before merging :)

We recently swapped our merge strategy to "Squash and merge" anyways so there is no need to squash the commits beforehand.

**Comment by tortmayr** [comment]
Thanks for the update. 
I guess we should also use an optional injectable class for the collision detection (similar to the routing improvements). This would also allow us to get rid of the "noElelementOverlap" option

**Comment by tortmayr** [comment]
@gabrieldulca  Thanks again for the improvements. I migrated the collision handling into an optional "MovementRestrictor" component and removed the now obsolote GLSPViewerOptions API.
@planger or @martin-fleck-at  Could you please review this latest change?
